nx_cugraph/drawing/layout.py

```diff
@@ -0,0 +1,151 @@
+    _dtype_param,
+    _get_float_dtype,
+    _seed_to_int,
+    networkx_algorithm,
+)
+
+__all__ = [
+    "forceatlas2_layout",
+]
+
+
+@networkx_algorithm(
+    extra_params={
+        "outbound_attraction_distribution : bool, default True": (
+            "Distributes attraction along outbound edges. "
+            "Hubs attract less and thus are pushed to the borders."
+        ),
+        **_dtype_param,
+    },
+    version_added="25.04",
+    _plc="forceatlas2_layout",
+)
+def forceatlas2_layout(
+    G,
+    pos=None,
+    *,
+    max_iter=100,
+    jitter_tolerance=1.0,
+    scaling_ratio=2.0,
+    gravity=1.0,
+    distributed_action=False,
+    strong_gravity=False,
+    node_mass=None,
+    node_size=None,
+    weight=None,
+    dissuade_hubs=False,
+    linlog=False,
+    seed=None,
+    dim=2,
+    store_pos_as=None,
+    # nx_cugraph-only argument
+    outbound_attraction_distribution=True,
+    dtype=None,
+):
+    """
+    `distributed_action`, `node_mass`, `node_size` parameter is currently ignored.
+    Only `dim=2` is supported. `outbound_attraction_distribution` is
+    a backend-only argument.
+    """
+    if len(G) == 0:
+        return {}
+
+    if dim != 2:
+        raise NotImplementedError(
+            f"dim={dim} not supported; only dim=2 is currently supported"
+        )
+
+    # Split dict into cupy arrays of XY coords for PLC
+    if pos is not None:
+        if not isinstance(pos, dict):
+            raise TypeError(f"pos must be dict or None; got {type(pos)}")
+
+        # NOTE currently only x & y (dim=2) coordinated are supported by PLC
+        #   greater dimensions should be supported in the future to align with nx
+        start_pos_arr = cp.array(list(pos.values()))
+        x_start = start_pos_arr[:, 0]
+        y_start = start_pos_arr[:, 1]
+
+    if weight is not None:
+        # match the float dtype on input graph's edgelist, default to float32
+        G = _to_graph(G, weight, 1, _get_float_dtype(dtype))
+        dtype = _get_float_dtype(dtype, graph=G, weight=weight)
+        G_plc = G._get_plc_graph(weight, 1, dtype)
+    else:
+        G = _to_graph(G)
+        G_plc = G._get_plc_graph()
+
+    seed = _seed_to_int(seed)
+
+    vertices, x_axis, y_axis = plc.force_atlas2(
+        plc.ResourceHandle(),
+        random_state=seed,
+        graph=G_plc,
+        max_iter=max_iter,
+        x_start=x_start,
+        y_start=y_start,
+        outbound_attraction_distribution=outbound_attraction_distribution,
+        lin_log_mode=linlog,
+        prevent_overlapping=dissuade_hubs,  # this might not be the right usage
+        jitter_tolerance=jitter_tolerance,
+        scaling_ratio=scaling_ratio,
+        strong_gravity_mode=strong_gravity,
+        gravity=gravity,
+    )
+
+    pos_arr = cp.vstack((x_axis, y_axis)).T
+    pos = {int(vertices[i]): cp.asnumpy(pos_arr[i]) for i in range(vertices.shape[0])}
+
+    if store_pos_as is not None:
+        nx.set_node_attributes(G, pos, store_pos_as)
```

Comment:
I like how clear the intent of this code is, but this won't work well today for a few reasons:
- `nx.set_node_attributes` is not yet dispatched (it's coming in nx 3.5)
  - see: https://github.com/networkx/networkx/pull/7824
- We don't implement it yet , but we probably should soon
- We should update the original `G` graph, not the converted graph

Handling different cases is probably a little annoying. See `relabel_nodes` as one example that sometimes modifies an input graph. We should at least handle when `G_orig` is an `nxcg.Graph` or `nxcg.CudaGraph` (this is all that `relabel_nodes` handles). I think FA2 could also handle an `nx.Graph` by calling `nx.set_node_attributes` as you currently do. `nxcg.CudaGraph` can simply save an array to `G.node_values` dict.

Reply:
Actually, to test this for development, I think you could comment out the `plc.force_atlas2` call and instead set e.g.
```python
vertices = cp.arange(len(G), dtype=index_dtype)
x_axis = cp.arange(len(G), dtype=dtype)
y_axis = cp.arange(len(G), dtype=dtype)
```